### PR TITLE
#423 Evaluator -- String substitution improvements

### LINKED ISSUE #423: Evaluator improvement -- stringSubstitution should be able to handle duplicate parameter substitutions
In other words, the following substitution:
```
{
 operator: "stringSubstitution",
 children: [
  "%1 is the same as %1 but not %2",
  "this",
  "that"
 ]
}
```
should return "this is the same as this but not that", but currently it returns "this is the same as that but not"

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fix #423 

Been meaning to fix this for a while, but now that it will be needed to solve #620 properly, have quickly done it.

Means that now a substitution parameter can be used more than once in a string, for example a url with a link:
```
To verify, please visit: [%1](%1)
```

Added a new test to handle this case. (Remember to use the "evaluator_test" snapshot to run the test suite)

### CODE REVIEW

File: src/generated/graphql.ts [hunk 1]
```diff
@@ -638,10 +638,14 @@ export type Application = Node & {
...
   /** Reads and enables pagination through a set of `Verification`. */
   verifications: VerificationsConnection;
-  /** Reads and enables pagination through a set of `OrganisationApplicationJoin`. */
-  organisationApplicationJoins: OrganisationApplicationJoinsConnection;
   /** Reads and enables pagination through a set of `UserApplicationJoin`. */
   userApplicationJoins: UserApplicationJoinsConnection;
+  /** Reads and enables pagination through a set of `OrganisationApplicationJoin`. */
+  organisationApplicationJoins: OrganisationApplicationJoinsConnection;
+  /** Reads and enables pagination through a set of `LicenceApplicationJoin`. */
+  licenceApplicationJoins: LicenceApplicationJoinsConnection;
```
Comment (nmadruga) [inline src/generated/graphql.ts:646]: A couple of additions on the graphQL generated file. You probably will need to run core_templates and re-commit this to clean up from other specific snapshots

**Review by nmadruga — CHANGES_REQUESTED** [review]
As mentioned in application-manager-app/issues/#620:

> * Nice if you can add this to all templates in `core_templates`
> * Add a note about this on the Template Builder docs - looks pretty empty now, but I think we need to start writing this small tips to help when we finally write the full on documentation for it.

File: src/generated/graphql.ts (hunk 1 above)
Comment (CarlosNZ) [reply src/generated/graphql.ts]: Yeah, I usually regenerate this file in develop after a significant merge

**Comment by CarlosNZ** [comment]
> Nice if you can add this to all templates in `core_templates`

Done! fc93370b
Note that these links won't show correctly until the published evaluator package is upgraded (which I'll do as soon as this is approved)

> Add a note about this on the Template Builder docs - looks pretty empty now, but I think we need to start writing this small tips to help when we finally write the full on documentation for it.

I don't think there's anything to document here -- we should never have been doing it that way to begin with. The natural thing to do would be to put links inside Markdown links. If you don't and you rely on native email client link parsing, you're asking for problems. But the problem is with the email clients, not our system, and the original problem was a very edge case even if you do it that way.

**Comment by nmadruga** [comment]
> > Add a note about this on the Template Builder docs - looks pretty empty now, but I think we need to start writing this small tips to help when we finally write the full on documentation for it.
> 
> I don't think there's anything to document here -- we should never have been doing it that way to begin with. The natural thing to do would be to put links inside Markdown links. If you don't and you rely on native email client link parsing, you're asking for problems. But the problem is with the email clients, not our system, and the original problem was a very edge case even if you do it that way.

Yes, my thinking was just when we do a hand-over to external users (which was asked) we have a road-map for then to follow. Documenting this kind of tip would really help us not having to deal with fixes required for problems that were just created by not using the specific required configuration. And there is no way to know if it's not written somewhere.

**Comment by CarlosNZ** [comment]
> Yes, my thinking was just when we do a hand-over to external users (which was asked) we have a road-map for then to follow. Documenting this kind of tip would really help us not having to deal with fixes required for problems that were just created by not using the specific required configuration. And there is no way to know if it's not written somewhere.

Yeah, the template builder docs are in the front-end anyway (and I can't actually see them anywhere tbh), so outside the scope of this issue I'd say. Why don't you add a quick note to the front-end and make that the PR for [this one](https://github.com/openmsupply/application-manager-server/issues/620)?

**Comment by nmadruga** [comment]
@CarlosNZ Can you add something to the Template Builder docs, please? I think just one line would be enough, is more as a note for us not to forget about adding this to the official docs sometime soon.

**Comment by CarlosNZ** [comment]
@nmadruga added a note: https://github.com/openmsupply/application-manager-web-app/pull/1050

**Review by nmadruga — APPROVED** [review]

**Pull request merged**